File: Template/Graph/Tarjan.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
#define endl '\n'

template<typename T>
class Tarjan {
public:
    /* 建圖*/
    int n;
    vector<vector<T>> g;

    /* Tarjan & BCC & v-BCC 相關 */
    vector<T> dfn, low;
    vector<T> stk;
    int time;
    set<pair<T, T>> bridges;
    vector<int> bcc_id;
    vector<vector<T>> bccs;
    vector<bool> cut_vertices;

    /* 圓方樹 (Block-Cut Tree) 相關 */
    int bct_sz;
    vector<vector<T>> bct;
    vector<T> weights;

    Tarjan(int n) : n(n) {
        g.resize(n);
        dfn.assign(n, -1);
        low.assign(n, -1);
        bcc_id.assign(n, -1);
        cut_vertices.assign(n, false);
        bct.resize(2 * n);
        weights.assign(2 * n, 0);
        for (int i = 0; i < n; i++) weights[i] = -1;
        bct_sz = n;
    }

    void add_edge(int u, int v) {
        g[u].push_back(v);
        g[v].push_back(u);
    }

    void add_bct_edge(int u, int v) {
        bct[u].push_back(v);
        bct[v].push_back(u);
    }

    void dfs1(int u, int fa) {
        dfn[u] = low[u] = time++;
        stk.push_back(u);
        int cnt = 0;
        for (auto v : g[u]) {
            if (v == fa) continue;
            if (dfn[v] != -1)
                low[u] = min(low[u], dfn[v]);
            else {
                cnt++;
                dfs1(v, u);
                low[u] = min(low[u], low[v]);
                if (low[v] > dfn[u])  // bridge
                    bridges.insert({min(u, v), max(u, v)});
                if (fa != -1 && low[v] >= dfn[u])  // cut vertex
                    cut_vertices[u] = true;
                if (low[v] >= dfn[u]) {  // v-BCC & Block-Cut Tree
                    int s = bct_sz++;
                    weights[s] = 1;
                    add_bct_edge(s, u);
                    int w;
                    do {
                        w = stk.back();
                        stk.pop_back();
                        add_bct_edge(s, w);
                        weights[s]++;
                    } while (w != v);
                }
            }
        }
        if (fa == -1 && cnt > 1)  // cut vertex of root
            cut_vertices[u] = true;
    }

    void dfs2(int u) {
        bcc_id[u] = bccs.size() - 1;
        bccs.back().push_back(u);
        for (auto v : g[u]) {
            if (bcc_id[v] != -1 || bridges.count({min(u, v), max(u, v)}))
                continue;
            dfs2(v);
        }
    }

    void run() {
        for (int u = 0; u < n; u++)
            if (dfn[u] == -1) dfs1(u, -1), stk.clear();
        for (int u = 0; u < n; u++)
            if (bcc_id[u] == -1) bccs.push_back(vector<T>()), dfs2(u);
    }
};

File: Template/Graph/Tarjan.py
class Tarjan:
    def __init__(self, n: int):
        self.n = n
        self.g = [[] for _ in range(n)]
        
        # --- Tarjan & BCC & v-BCC 相關 ---
        self.dfn = [-1] * n
        self.low = [-1] * n
        self.stk = []
        self.time = 0
        self.bridges = set()
        self.bcc_id = [-1] * n
        self.bccs = []
        self.cut_vertices = [False] * n

        # --- 圓方樹 (Block-Cut Tree) 相關 ---
        self.bct_sz = n
        self.bct = [[] for _ in range(2 * n)]
        self.weights = [-1] * n + [0] * n

    def add_edge(self, u: int, v: int):
        self.g[u].append(v)
        self.g[v].append(u)

    def add_bct_edge(self, u: int, v: int):
        self.bct[u].append(v)
        self.bct[v].append(u)
    
    def dfs1(self, u: int, fa: int):
        """遞迴版本"""
        # self.dfn[u] = self.low[u] = self.time
        # self.time += 1
        # self.stk.append(u)
        # cnt = 0
        # for v in self.g[u]:
        #     if v == fa:
        #         continue
        #     if self.dfn[v] != -1:
        #         self.low[u] = min(self.low[u], self.dfn[v])
        #     else:
        #         cnt += 1
        #         self.dfs1(v, u)
        #         self.low[u] = min(self.low[u], self.low[v])
        #         if self.low[v] > self.dfn[u]:  # bridge
        #             self.bridges.add((min(u, v), max(u, v)))
        #         if fa != -1 and self.low[v] >= self.dfn[u]:  # cut vertex
        #             self.cut_vertices[u] = True
        #         if self.low[v] >= self.dfn[u]:  # v-BCC & Block-Cut Tree
        #             s = self.bct_sz
        #             self.bct_sz += 1
                    
        #             # 從棧中彈出節點，它們都屬於這個 v-BCC
        #             vbcc_sz = 1
        #             self.add_bct_edge(s, u)
        #             while True:
        #                 w = self.stk.pop()
        #                 self.add_bct_edge(s, w)
        #                 vbcc_sz += 1
        #                 if w == v:
        #                     break
                    
        #             # 方點的權重是 v-BCC 的大小
        #             self.weights[s] = vbcc_sz
        # if fa == -1 and cnt > 1:  # cut vertex of root
        #     self.cut_vertices[u] = True
        """迭代版本"""
        st = [(u, fa, 0, 0)]  # (u, fa, i, cnt)
        while st:
            u, fa, i, cnt = st.pop()
            if i == 0:
                self.dfn[u] = self.low[u] = self.time
                self.time += 1
                self.stk.append(u)
            if i > 0:  # 處理 i - 1 的返回邏輯
                v = self.g[u][i-1]
                self.low[u] = min(self.low[u], self.low[v])
                if self.low[v] > self.dfn[u]:  # bridge
                    self.bridges.add((min(u, v), max(u, v)))
                if fa != -1 and self.low[v] >= self.dfn[u]:  # cut vertex
                    self.cut_vertices[u] = True
                if self.low[v] >= self.dfn[u]:  # v-BCC & Block-Cut Tree
                    s = self.bct_sz
                    self.bct_sz += 1
                    # 從棧中彈出節點，它們都屬於這個 v-BCC
                    vbcc_sz = 1
                    self.add_bct_edge(s, u)
                    while True:
                        w = self.stk.pop()
                        self.add_bct_edge(s, w)
                        vbcc_sz += 1
                        if w == v:
                            break
                    # 方點的權重是 v-BCC 的大小
                    self.weights[s] = vbcc_sz
            for j in range(i, len(self.g[u])):
                v = self.g[u][j]
                if v == fa:
                    continue
                if self.dfn[v] != -1:
                    self.low[u] = min(self.low[u], self.dfn[v])
                else:
                    st.append((u, fa, j + 1, cnt + 1))
                    st.append((v, u, 0, 0))
                    break
            else:
                if fa == -1 and cnt > 1:  # cut vertex of root
                    self.cut_vertices[u] = True

    def dfs2(self, u: int):
        """遞迴版本"""
        # self.comp_id[u] = len(self.bccs) - 1
        # self.bccs[-1].append(u)
        # for v in self.g[u]:
        #     if self.comp_id[v] != -1 or (min(u, v), max(u, v)) in self.bridges:
        #         continue
        #     self.dfs2(v)
        """迭代版本"""
        st = [u]
        while st:
            u = st.pop()
            self.bcc_id[u] = len(self.bccs) - 1
            self.bccs[-1].append(u)
            for v in self.g[u]:
                if self.bcc_id[v] != -1 or (min(u, v), max(u, v)) in self.bridges:
                    continue
                st.append(v)

    def run(self):
        for u in range(self.n):
            if self.dfn[u] == -1:
                self.dfs1(u, -1)
                self.stk.clear()
        for u in range(self.n):
            if self.bcc_id[u] == -1:
                self.bccs.append([])
                self.dfs2(u)